sdk/storage_datalake/src/authorization_policies/auto_refreshing_token.rs

```diff
@@ -0,0 +1,65 @@
+use async_lock::RwLock;
+use azure_core::{
+    auth::{TokenCredential, TokenResponse},
+    Error,
+};
+use chrono::{Duration, Utc};
+use std::sync::Arc;
+
+fn is_expired(token: TokenResponse) -> bool {
+    token.expires_on < Utc::now() + Duration::seconds(20)
+}
+
+#[derive(Clone)]
+pub struct AutoRefreshingTokenCredential {
+    credential: Arc<dyn TokenCredential>,
+    current_token: Arc<RwLock<Option<std::result::Result<TokenResponse, Error>>>>,
+}
+
+impl std::fmt::Debug for AutoRefreshingTokenCredential {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.debug_struct("AutoRefreshingTokenCredential")
+            .field("credential", &"TokenCredential")
+            .finish()
+    }
+}
+
+impl AutoRefreshingTokenCredential {
```

Comment:
Shouldn't this go in `azure_core`?

Reply:
It will eventually. I know @roeap wanted to do it in stages.

Reply:
I completely agree :). If we can move this there now, great. Similarly, the pipeline authorization policies should likely end up at least in `azure_storage`, or if they are useful in other crates as well maybe also core?

If this is OK, I am happy to move this to core in this PR.

Reply:
Moved the `AutoRefreshingCredential` into the identity crate to the other `TokenCredentials`. Was considering to also move the policies to the `storage_core` crate, but thought it better to do this when migrating blob to the pipeline architecture.